### ISSUE TITLE
[Phase 4] Probability Calibration

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/clinical-risk-stratification-engine#5]
## Summary

Implement isotonic regression calibration to ensure probability estimates are reliable for clinical decision-making. Uncalibrated models often produce overconfident or underconfident predictions.

**Parent Issue:** #1
**Depends On:** #4 (Phase 3)

## Objectives

- Calibrate probability outputs for each outcome
- Measure calibration quality (Brier score, reliability diagrams)
- Ensure calibration generalizes to new data

## Files to Create/Modify

| File | Action | Description |
|------|--------|-------------|
| `src/calibration/isotonic.py` | Create | Isotonic calibration wrapper |
| `src/calibration/metrics.py` | Create | Brier score, reliability diagrams |
| `tests/test_calibration.py` | Create | Calibration tests |

### Isotonic Calibration
- [ ] Implement wrapper around sklearn's CalibratedClassifierCV
- [ ] Support fitting on held-out calibration set
- [ ] Store calibration mappings for each output

### Calibration Metrics
- [ ] Implement Brier score calculation
- [ ] Implement Expected Calibration Error (ECE)
- [ ] Generate reliability diagrams
- [ ] Support multi-output comparison

### Integration
- [ ] Ensure calibration uses held-out data (not training data)
- [ ] Combine with multi-output model

### Isotonic Calibrator

```python
# src/calibration/isotonic.py

from sklearn.calibration import CalibratedClassifierCV
from sklearn.isotonic import IsotonicRegression
import numpy as np


class MultiOutputCalibrator:
    """
    Post-hoc probability calibration for multi-output classifier.

    Learning Context:
    - Raw model probabilities are often poorly calibrated
    - Isotonic regression is a non-parametric method that learns a monotonic mapping
    - Calibration must be done on HELD-OUT data (not training data)
    - After calibration, P(y=1|model says 0.7) should actually be ~70%
    """

    def __init__(self, method='isotonic'):
        self.method = method
        self.calibrators_ = {}
        self.target_names = ['readmission', 'sepsis', 'mortality']

    def fit(self, y_true: dict, y_prob: dict):
        """
        Fit calibrators on held-out calibration data.

        Parameters
        ----------
        y_true : dict
            True labels for each target {target_name: array of 0/1}
        y_prob : dict
            Predicted probabilities {target_name: array of probabilities}
        """
        for target in self.target_names:
            if target in y_true and target in y_prob:
                calibrator = IsotonicRegression(out_of_bounds='clip')
                # Fit on probability of positive class
                calibrator.fit(y_prob[target], y_true[target])
                self.calibrators_[target] = calibrator

        return self

    def calibrate(self, y_prob: dict) -> dict:
        """
        Apply calibration to new predictions.

        Parameters
        ----------
        y_prob : dict
            Uncalibrated probabilities {target_name: array}

        Returns
        -------
        dict
            Calibrated probabilities
        """
        calibrated = {}
        for target, probs in y_prob.items():
            if target in self.calibrators_:
                calibrated[target] = self.calibrators_[target].predict(probs)
            else:
                calibrated[target] = probs
        return calibrated


def calibrate_with_cv(model, X, y, cv=5, method='isotonic'):
    """
    Calibrate model using cross-validation to avoid overfitting.

    This is preferred when you don't have a dedicated calibration set.
    """
    return CalibratedClassifierCV(model, method=method, cv=cv)
```

### Calibration Metrics

```python
# src/calibration/metrics.py

import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import brier_score_loss
from sklearn.calibration import calibration_curve


def brier_score_multi(y_true: dict, y_prob: dict) -> dict:
    """
    Compute Brier score for each target.

    Learning Context:
    - Brier score = mean squared error of probability predictions
    - Range: 0 (perfect) to 1 (worst)
    - Brier score < 0.25 is generally considered good
    - Calibration should REDUCE Brier score
    """
    scores = {}
    for target in y_true.keys():
        if target in y_prob:
            scores[target] = brier_score_loss(y_true[target], y_prob[target])
    return scores


def expected_calibration_error(y_true, y_prob, n_bins=10):
    """
    Compute Expected Calibration Error (ECE).

    ECE measures the average gap between predicted probability and
    actual frequency across probability bins.

    Lower is better. ECE < 0.05 is well-calibrated.
    """
    bin_boundaries = np.linspace(0, 1, n_bins + 1)
    ece = 0.0

    for i in range(n_bins):
        in_bin = (y_prob >= bin_boundaries[i]) & (y_prob < bin_boundaries[i + 1])
        if in_bin.sum() > 0:
            bin_accuracy = y_true[in_bin].mean()
            bin_confidence = y_prob[in_bin].mean()
            bin_weight = in_bin.sum() / len(y_true)
            ece += bin_weight * abs(bin_accuracy - bin_confidence)

    return ece


def plot_reliability_diagram(y_true, y_prob, title='Reliability Diagram', ax=None):
    """
    Plot reliability diagram showing calibration quality.

    A well-calibrated model should have points along the diagonal.
    Points above diagonal = underconfident
    Points below diagonal = overconfident
    """
    if ax is None:
        fig, ax = plt.subplots(figsize=(8, 6))

    prob_true, prob_pred = calibration_curve(y_true, y_prob, n_bins=10)

    ax.plot([0, 1], [0, 1], 'k--', label='Perfectly calibrated')
    ax.plot(prob_pred, prob_true, 's-', label='Model')
    ax.set_xlabel('Mean predicted probability')
    ax.set_ylabel('Fraction of positives')
    ax.set_title(title)
    ax.legend()
    ax.grid(True, alpha=0.3)

    return ax


def plot_multi_reliability(y_true: dict, y_prob: dict, y_prob_calibrated: dict = None):
    """Plot reliability diagrams for all targets, before/after calibration."""
    n_targets = len(y_true)
    fig, axes = plt.subplots(1, n_targets, figsize=(5 * n_targets, 5))

    if n_targets == 1:
        axes = [axes]

    for ax, target in zip(axes, y_true.keys()):
        plot_reliability_diagram(y_true[target], y_prob[target], f'{target} (uncalibrated)', ax)
        if y_prob_calibrated and target in y_prob_calibrated:
            prob_true_cal, prob_pred_cal = calibration_curve(
                y_true[target], y_prob_calibrated[target], n_bins=10
            )
            ax.plot(prob_pred_cal, prob_true_cal, 'o-', label='Calibrated')
            ax.legend()

    plt.tight_layout()
    return fig
```

## Verification

```bash
# Run tests
pytest tests/test_calibration.py -v

# Quick verification
python -c "
from src.calibration.metrics import brier_score_multi, expected_calibration_error
import numpy as np

# Simulated predictions
y_true = {'readmission': np.array([0, 0, 1, 0, 1])}
y_prob = {'readmission': np.array([0.1, 0.3, 0.8, 0.2, 0.6])}

brier = brier_score_multi(y_true, y_prob)
print(f'Brier score: {brier}')
"
```

## Definition of Done

- [ ] Isotonic calibration fits without errors
- [ ] Calibrated probabilities improve Brier score
- [ ] Reliability diagrams show improved calibration
- [ ] ECE decreases after calibration
- [ ] Calibration works for all 3 outputs
- [ ] All unit tests pass

## Technical Notes

**Learning Context:**
- Understand the difference between accuracy and calibration
- Learn why clinical decisions need calibrated probabilities
- Study isotonic vs Platt scaling tradeoffs

**Potential Challenges:**
- Calibration may be unstable with small minority class
- Need sufficient calibration set size (100+ per class recommended)
